tremor-script/src/ast/deploy/raw.rs

```diff
@@ -0,0 +1,785 @@
+            mid: helper.add_meta_w_name(self.start, self.end, &self.id),
+            module: helper.module.clone(),
+            id: self.id,
+            spec: spec.clone(),
+            args,
+            links,
+        };
+        helper.module.pop();
+        let script_name = flow_decl.fqsn(&helper.module);
+        helper.flow_defns.insert(script_name, flow_decl.clone());
+        Ok(flow_decl)
+    }
+}
+
+pub type DeployLinksRaw<'script> = Vec<DeployLinkRaw<'script>>;
+
+/// we're forced to make this pub because of lalrpop
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct DeployLinkRaw<'script> {
+    /// we're forced to make this pub because of lalrpop
+    pub from: DeployEndpointRaw<'script>,
+    /// we're forced to make this pub because of lalrpop
+    pub to: DeployEndpointRaw<'script>,
+}
+
+/// we're forced to make this pub because of lalrpop
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct CreateStmtRaw<'script> {
+    pub(crate) start: Location,
+    pub(crate) end: Location,
+    pub(crate) id: IdentRaw<'script>,
+    pub(crate) params: Option<DeployWithExprsRaw<'script>>,
+    /// Id of the definition
+    pub target: IdentRaw<'script>,
+    /// Module of the definition
+    pub(crate) module: Vec<IdentRaw<'script>>,
+}
+impl_expr!(CreateStmtRaw);
+
+/// A fully resolved deployable artefact
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub enum AtomOfDeployment<'script> {
+    /// A deployable pipeline instance
+    Pipeline(PipelineDecl<'script>),
+    /// A deployable connector instance
+    Connector(ConnectorDecl<'script>),
+    /// A deployable flow instance
+    Flow(FlowDecl<'script>),
+}
+
+impl<'script> Upable<'script> for CreateStmtRaw<'script> {
+    type Target = CreateStmt<'script>;
+    fn up<'registry>(self, helper: &mut Helper<'script, 'registry>) -> Result<Self::Target> {
+        // NOTE As we can have module aliases and/or nested modules within script definitions
+        // that are private to or inline with the script - multiple script definitions in the
+        // same module scope can share the same relative function/const module paths.
+        //
+        // We add the script name to the scope as a means to distinguish these orthogonal
+        // definitions. This is achieved with the push/pop pointcut around the up() call
+        // below. The actual function registration occurs in the up() call in the usual way.
+        //
+        helper.module.push(self.id.to_string());
+
+        // TODO check that names across pipeline/flow/connector definitions are unique or else hygienic error
+
+        let target_module = self
+            .module
+            .iter()
+            .map(|x| format!("{}::", x.to_string()))
+            .collect::<Vec<String>>()
+            .join("");
+        let fqsn = format!("{}{}", target_module, self.target.to_string());
+
+        let mut h2 = helper.clone(); // Clone to save some accounting/tracking & cleanup
+        h2.module = vec![];
```

Comment:
Is it save to remove all modules? Wouldn't that break modularity when used with `use` and `mod`?

Reply:
This is sneaky - we're removing modules from a clone of helper - so this gives us a fake `root` - and the empty modules is effectively the same as loading these from a file. This is sneaky enough that, I fear, it needs a lot more test cases. This is sadly consistent with other module management sneakiness where we inject a fake module to avoid collisions.

As we now have local, core ( special case ) absolute and relative module namespacing and resolution where we only had local, core and absolute before - i suspect we need to revisit this and perhaps other helper sneakiness across the AST.

Reply:
I feel guiltier ( yet shockingly productive! ) with making Helper cloneable - it really shouldn't be...

Reply:
I think the refactoring for helper to be non-clone fixes this.